ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionContractKeysValidation.scala

```diff
@@ -0,0 +1,295 @@
+import com.daml.lf.value.Value.ContractId
+
+object TransactionContractKeysValidation {
+
+  private type RawContractId = String
+
+  private[transaction] def validate(
+      transactionCommitter: TransactionCommitter
+  ): Step[DamlTransactionEntrySummary] =
+    (commitContext, transactionEntry) => {
+      val damlState = commitContext
+        .collectInputs[(DamlStateKey, DamlStateValue), Map[DamlStateKey, DamlStateValue]] {
+          case (key, Some(value)) if key.hasContractKey => key -> value
+        }
+      val contractKeyDamlStateKeysToContractIds: Map[DamlStateKey, RawContractId] =
+        damlState.collect {
+          case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty =>
+            k -> v.getContractKeyState.getContractId
+        }
+      val contractKeyDamlStateKeys: Set[DamlStateKey] =
+        contractKeyDamlStateKeysToContractIds.keySet
+      val contractKeysToContractIds: Map[DamlContractKey, RawContractId] =
+        contractKeyDamlStateKeysToContractIds.map(m => m._1.getContractKey -> m._2)
+
+      validateTransactionContractKeysUniquenessAndConsistency(
+        transactionCommitter,
+        commitContext.recordTime,
+        transactionEntry,
+        contractKeyDamlStateKeys,
+        contractKeysToContractIds,
+      ) match {
+        case StepContinue(transactionEntry) =>
+          validateTransactionContractKeysCausalMonotonicity(
+            transactionCommitter,
+            commitContext.recordTime,
+            transactionEntry,
+            contractKeyDamlStateKeys,
+            damlState,
+          )
+        case err => err
+      }
+    }
+
+  /** LookupByKey nodes themselves don't actually fetch the contract.
+    * Therefore we need to do an additional check on all contract keys
+    * to ensure the referred contract satisfies the causal monotonicity invariant.
+    * This could be reduced to only validate this for keys referred to by
+    * NodeLookupByKey.
+    */
+  private def validateTransactionContractKeysCausalMonotonicity(
+      transactionCommitter: TransactionCommitter,
+      recordTime: Option[Timestamp],
+      transactionEntry: DamlTransactionEntrySummary,
+      keys: Set[DamlStateKey],
+      damlState: Map[DamlStateKey, DamlStateValue],
+  ): StepResult[DamlTransactionEntrySummary] = {
+    val causalKeyMonotonicity = keys.forall { key =>
+      val state = damlState(key)
+      val keyActiveAt =
+        Conversions.parseTimestamp(state.getContractKeyState.getActiveAt).toInstant
+      !keyActiveAt.isAfter(transactionEntry.ledgerEffectiveTime.toInstant)
+    }
+    if (causalKeyMonotonicity)
+      StepContinue(transactionEntry)
+    else
+      transactionCommitter.reject(
+        recordTime,
+        transactionCommitter.buildRejectionLogEntry(
+          transactionEntry,
+          RejectionReason.Inconsistent("Causal monotonicity violated"),
+        ),
+      )
+  }
+
+  private sealed abstract class KeyValidationError()
```

Comment:
Please move these classes to the bottom of the file.

Reply:
Done.